File: SudokuClient/model/decisionTree.ts
import {Board} from './board';
import {Decision} from './decision'

export class DecisionTree {
    board: Board;
    root: Decision;
    currentDecision: Decision;
    decisionLog = new Array<string>();

    constructor(_board: Board)
    {
        this.board = _board;
    }
    
    public reset()
    {
        while(this.currentDecision != null)
        {
            this.currentDecision.reset();
            this.currentDecision = this.currentDecision.parent;
        }
        this.currentDecision = this.root;
    }

    public undoLastDecision()
    {
        if(this.currentDecision != null)
        {
            this.currentDecision.reset();
            this.currentDecision = this.currentDecision.parent;
        }
    }

    public nextDecision()
    {
        this.currentDecision = new Decision(this.board.getFirstUndefinedCell(), this.currentDecision);
        if (this.root == null)
            this.root = this.currentDecision;
        this.currentDecision.perform();
        this.decisionLog.push("N " + this.currentDecision.cell.toString());
        return this.currentDecision;
    }

    public changeDecision()
    {
        this.currentDecision = this.currentDecision.change();
        this.decisionLog.push("C " + this.currentDecision.cell.toString());
        return this.currentDecision;
    }

    public isForBoard(_board: Board)
    {
        return this.board == _board;
    }
    
}

File: SudokuServer/SudokuClient/sudoku/DecisionTree.cs
using System;
using System.Collections;

namespace Sudoku_Core
{
	/// <summary>
	/// Summary description for DecisionTree.
	/// </summary>
	public class DecisionTree
	{
		Board board;
		Decision root;
		Decision currentDecision;
		ArrayList decisionLog = new ArrayList();

		public DecisionTree(Board _board)
		{
			board = _board;
//			Cell cell = board.GetFirstUndefinedCell();
//			root = new Decision(cell);
//			currentDecision = root;
		}

		public void Reset()
		{
			while(currentDecision != null)
			{
				currentDecision.Reset();
				currentDecision = currentDecision.Parent;
			}
			currentDecision = root;
		}

		public void UndoLastDecision()
		{
			if(currentDecision != null)
			{
				currentDecision.Reset();
				currentDecision = currentDecision.Parent;
			}
		}

		public Decision NextDecision()
		{
			currentDecision = new Decision(board.GetFirstUndefinedCell(), currentDecision);
			if (root == null)
				root = currentDecision;
			currentDecision.Perform();
			decisionLog.Add("N " + currentDecision.Cell.ToString());
			return currentDecision;
		}

		public Decision ChangeDecision()
		{
			currentDecision = currentDecision.Change();
			decisionLog.Add("C " + currentDecision.Cell.ToString());
			return currentDecision;
		}

		public Decision Root
		{
			get {return root;}
		}

		public bool IsForBoard(Board _board)
		{
			return board == _board;
		}
	}
}
